src/node_modules/logger.js

```diff
@@ -187,31 +191,35 @@ export default class Logger extends Writable {
 
 			connection.setTimeout(opts.inactivityTimeout || defaults.inactivityTimeout);
 			// Expose the connection’s drain event
-			connection.on('drain', () => {
-				setImmediate(() => this.emit('connection drain'));
-			});
 
-			connection.on('connect', () => {
-				this.emit('connected');
-			});
+			return connection;
+		});
 
-			connection.on('disconnect', () => {
-				this.emit('disconnected')
-			});
+		this.on('drain', () => {
+			this[$bufferWasFull] = false;
+		});
 
-			return connection;
+		this.ringBuffer.on('buffer shift', () => {
+			if (!this[$bufferWasFull]) {
+				this.debugLogger.log("Buffer is full, will be shifting records until buffer is drained.");
+				this[$bufferWasFull] = true;
+			}
 		});
 	}
 
 	// Stream methods
 
 	@unwritable
-	async _write(log, enc, cb) {
-		log = finalizeLogString(log, this[$token]);
-
+	async _write(ch, enc, cb) {
 		try {
 			const conn = await this[$getConnection]();
-			conn.write(log);
+			var record = this.ringBuffer.read();
+			if (record != null) {
```

Comment:
`_write` is called whenever `write()` of `Writable` is called so there will always be some data in the array backing ringbuffer when we call `read()`.